File: python/src/fancy_expr/_tokenize.py
"""The tokenizer.

Hand-written, like the other two implementations, and for the same reason: the
three standard libraries in this space disagree with each other on the semantics
in ``GRAMMAR.md``, so adopting one would ship a fourth dialect rather than a
third implementation of one.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

TokenKind = Literal["number", "string", "ident", "punct", "eof"]

# Longest-first. `===` must be tried before `==`, and `!==` before `!=` and `!`,
# or the tokenizer splits a valid operator into two invalid ones and the error
# points at the wrong offset.
_PUNCT: tuple[str, ...] = (
    "===",
    "!==",
    "==",
    "!=",
    "&&",
    "||",
    "<=",
    ">=",
    "?",
    ":",
    ".",
    ",",
    "(",
    ")",
    "[",
    "]",
    "{",
    "}",
    "+",
    "-",
    "*",
    "/",
    "<",
    ">",
    "!",
)

_IDENT_START = set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_$")
_IDENT_REST = _IDENT_START | set("0123456789")
_DIGITS = set("0123456789")


class ExprSyntaxError(ValueError):
    """A malformed expression.

    NEVER conflate this with a ``None`` result. An unresolved path is ``None``
    and means *absent*; this means the author wrote something the grammar cannot
    read. Collapsing the two is the defect this package exists to remove -- a
    branch condition that could not be evaluated read as ``False`` and a live
    workflow took the wrong road on every run while reporting success.
    """

    def __init__(self, message: str, offset: int, expression: str) -> None:
        super().__init__(f"{message} (at offset {offset} in {expression!r})")
        self.offset = offset
        self.expression = expression


@dataclass(frozen=True, slots=True)
class Token:
    kind: TokenKind
    value: Any
    start: int


def tokenize(expression: str) -> list[Token]:
    tokens: list[Token] = []
    i = 0
    n = len(expression)

    while i < n:
        ch = expression[i]

        if ch in " \t\r\n":
            i += 1
            continue

        if ch in ("'", '"'):
            start = i
            text, i = _read_string(expression, i)
            tokens.append(Token("string", text, start))
            continue

        # A leading `.` on a number is NOT accepted: `.5` would be ambiguous
        # with the member operator, and every path segment begins with an
        # identifier character.
        if ch in _DIGITS:
            start = i
            number, i = _read_number(expression, i)
            tokens.append(Token("number", number, start))
            continue

        if ch in _IDENT_START:
            start = i
            while i < n and expression[i] in _IDENT_REST:
                i += 1
            tokens.append(Token("ident", expression[start:i], start))
            continue

        for p in _PUNCT:
            if expression.startswith(p, i):
                tokens.append(Token("punct", p, i))
                i += len(p)
                break
        else:
            raise ExprSyntaxError(f"Unexpected character {ch!r}", i, expression)

    tokens.append(Token("eof", None, n))
    return tokens


def _read_string(expression: str, i: int) -> tuple[str, int]:
    quote = expression[i]
    start = i
    i += 1
    out: list[str] = []

    while i < len(expression):
        ch = expression[i]

        if ch == "\\":
            if i + 1 >= len(expression):
                raise ExprSyntaxError("Unterminated escape", i, expression)
            nxt = expression[i + 1]
            out.append(
                {"n": "\n", "t": "\t", "r": "\r", "\\": "\\", "'": "'", '"': '"'}.get(nxt, nxt)
            )
            i += 2
            continue

        if ch == quote:
            return "".join(out), i + 1

        out.append(ch)
        i += 1

    # Row 0801. An unclosed string is the most common malformed expression there
    # is, and it MUST reach the caller as a failure rather than as a value.
    raise ExprSyntaxError("Unterminated string", start, expression)


def _read_number(expression: str, i: int) -> tuple[float | int, int]:
    start = i
    n = len(expression)

    while i < n and expression[i] in _DIGITS:
        i += 1

    is_float = False

    if i < n and expression[i] == "." and i + 1 < n and expression[i + 1] in _DIGITS:
        is_float = True
        i += 1
        while i < n and expression[i] in _DIGITS:
            i += 1

    if i < n and expression[i] in "eE":
        j = i + 1
        if j < n and expression[j] in "+-":
            j += 1
        if j < n and expression[j] in _DIGITS:
            is_float = True
            i = j
            while i < n and expression[i] in _DIGITS:
                i += 1

    text = expression[start:i]

    # An integral literal stays an `int`. JSON has one number type and so does
    # this grammar, but Python prints `4` and `4.0` differently and the goldens
    # are written the way JSON writes them -- `in.count + 1` is `4`, not `4.0`.
    return (float(text) if is_float else int(text)), i

File: src/tokenize.ts
/**
 * The tokenizer. Turns source text into a flat token list, or throws.
 *
 * Kept separate from the parser so a malformed *character* (an unterminated
 * string, a stray `@`) fails here with an offset, rather than surfacing as a
 * confusing parse error about something three tokens later.
 */

export type TokenType =
  | "number"
  | "string"
  | "ident"
  | "punct"
  | "eof";

export type Token = {
  type: TokenType;
  /** For `number` the numeric value; for everything else the source text. */
  value: string | number;
  /** Byte offset of the token's first character, for error messages. */
  at: number;
};

/** Thrown for anything that cannot be tokenized or parsed. NEVER returned. */
export class ExprSyntaxError extends Error {
  constructor(
    message: string,
    readonly expression: string,
    readonly at: number,
  ) {
    super(`${message} (at offset ${at} in \`${expression}\`)`);
    this.name = "ExprSyntaxError";
  }
}

/**
 * Multi-character operators, longest first.
 *
 * Order matters and is the classic tokenizer trap: match `=` before `===` and
 * every strict comparison silently becomes two tokens. Sorted by length at
 * module load so adding one cannot reintroduce that.
 */
const PUNCT = [
  "===", "!==", "==", "!=", "<=", ">=", "&&", "||",
  "?", ":", ".", ",", "(", ")", "[", "]", "{", "}", "+", "-", "*", "/", "<", ">", "!",
].sort((a, b) => b.length - a.length);

const IDENT_START = /[A-Za-z_$]/;
const IDENT_REST = /[A-Za-z0-9_$-]/;

export function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  while (i < source.length) {
    const ch = source[i];

    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
      i += 1;
      continue;
    }

    // Strings. Both quote styles, because authors use both, with backslash
    // escapes for the quote itself and the backslash.
    if (ch === "'" || ch === '"') {
      const start = i;
      const quote = ch;
      let out = "";
      i += 1;
      let closed = false;

      while (i < source.length) {
        if (source[i] === "\\" && i + 1 < source.length) {
          out += source[i + 1];
          i += 2;
          continue;
        }
        if (source[i] === quote) {
          i += 1;
          closed = true;
          break;
        }
        out += source[i];
        i += 1;
      }

      if (!closed) throw new ExprSyntaxError("Unterminated string", source, start);
      tokens.push({ type: "string", value: out, at: start });
      continue;
    }

    // Numbers. Digits only at the start — a leading `-` is unary minus and is
    // the parser's business, so `1-2` tokenizes as three tokens rather than two.
    if (ch >= "0" && ch <= "9") {
      const start = i;
      while (i < source.length && source[i] >= "0" && source[i] <= "9") i += 1;
      if (source[i] === "." && source[i + 1] >= "0" && source[i + 1] <= "9") {
        i += 1;
        while (i < source.length && source[i] >= "0" && source[i] <= "9") i += 1;
      }
      tokens.push({ type: "number", value: Number(source.slice(start, i)), at: start });
      continue;
    }

    if (IDENT_START.test(ch)) {
      const start = i;
      i += 1;
      while (i < source.length && IDENT_REST.test(source[i])) i += 1;
      tokens.push({ type: "ident", value: source.slice(start, i), at: start });
      continue;
    }

    const punct = PUNCT.find((p) => source.startsWith(p, i));
    if (punct) {
      tokens.push({ type: "punct", value: punct, at: i });
      i += punct.length;
      continue;
    }

    throw new ExprSyntaxError(`Unexpected character \`${ch}\``, source, i);
  }

  tokens.push({ type: "eof", value: "", at: source.length });
  return tokens;
}
